ros2topic/ros2topic/verb/echo.py

```diff
@@ -0,0 +1,148 @@
+
+
+def main(args):
+    if not args.csv:
+        register_yaml_representer()
+        callback = subscriber_cb
+    else:
+        callback = subscriber_cb_csv
+    with DirectNode(args) as node:
+        subscriber(node, args.topic_name, args.message_type, callback)
+
+
+def register_yaml_representer():
+    # Register our custom representer for YAML output
+    yaml.add_representer(OrderedDict, represent_ordereddict)
+
+
+# Custom representer for getting clean YAML output that preserves the order in
+# an OrderedDict.
+# Inspired by:
+# http://stackoverflow.com/a/16782282/7169408
+def represent_ordereddict(dumper, data):
+    items = []
+    for k, v in data.items():
+        items.append((dumper.represent_data(k), dumper.represent_data(v)))
+    return yaml.nodes.MappingNode(u'tag:yaml.org,2002:map', items)
+
+
+def subscriber(node, topic_name, message_type, callback):
+    if message_type is None:
+        topic_names_and_types = get_topic_names_and_types(node=node)
+        for n, t in topic_names_and_types:
+            if n == topic_name:
+                message_type = t
+                break
+        else:
+            raise RuntimeError(
+                'Could not determine the type for the passed topic')
+
+    try:
+        package_name, message_name = message_type.split('/', 2)
+    except ValueError:
+        raise RuntimeError('The passed message type is invalid')
+    module = importlib.import_module(package_name + '.msg')
+    msg_module = getattr(module, message_name)
+
+    node.create_subscription(
+        msg_module, topic_name, callback, qos_profile=qos_profile_sensor_data)
+
+    while rclpy.ok():
+        rclpy.spin_once(node)
+
+
+def subscriber_cb(msg):
+    print(msg_to_yaml(msg))
+
+
+def msg_to_yaml(msg):
+    return yaml.dump(msg_to_ordereddict(msg), width=sys.maxsize)
+
+
+def subscriber_cb_csv(msg):
+    print(msg_to_csv(msg))
+
+
+def msg_to_csv(msg):
+    def to_string(val):
+        r = ''
+        if any([isinstance(val, t) for t in [list, tuple]]):
+            for v in val:
+                if r:
+                    r += ','
+                r += to_string(v)
+        elif any([isinstance(val, t) for t in [bool, bytes, float, int, str]]):
+            r = str(val)
+        else:
+            r = msg_to_csv(val)
+        return r
+    result = ''
+    # We rely on __slots__ retaining the order of the fields in the .msg file.
+    for field_name in msg.__slots__:
+        value = getattr(msg, field_name, None)
+        if result:
+            result += ','
+        result += to_string(value)
+    return result
+
+
+# Convert a msg to an OrderedDict. We do this instead of implementing a generic
+# __dict__() method in the msg because we want to preserve order of fields from
+# the .msg file(s).
+def msg_to_ordereddict(msg):
+    d = OrderedDict()
+    # We rely on __slots__ retaining the order of the fields in the .msg file.
+    types = [bool, bytes, dict, float, int, list, str, tuple, OrderedDict]
```

Comment:
It looks that was needed by @gerkey in https://github.com/ros2/cli_tools/pull/5 but I don't know why. It doesn't look to be used for recursion because it's listed as part of the `types` list that I understood as "primitive types that don't need recursion". Both the slots field and every array field in the python messages are represented as list so I don't see what the OrderedDict refers to